### ISSUE TITLE
GL Surface sometimes render as all black, sometimes crashes app

### ISSUE THREAD

**Issue by dcworldwide** [dcworldwide/react-native-gl-image-filter-examples#1]

**Comment by st0ffern** [comment]
@dcworldwide would this be of any interest? 
https://github.com/stoffern/gl-react-instagramfilters

**Comment by dcworldwide** [comment]
@stoffern Bravo :) Nice to see someone has solved the same issue. Out of interest, in my lib, I've had issues with the GL surface or GL image crashing / rendering a black screen. Have you experienced any such behavior? The issue occurs in conjunction to loading a image captured via RN Camera. Interestingly, the problem seems to relate to loading large images, as smaller files have worked fine.

**Comment by st0ffern** [comment]
nope no problems with that, i have not tested all filters tho, only Amaro i and it worked fine.

**Comment by dcworldwide** [comment]
Ok. I'll get around to testing out the other filters in your repo in a week or so.

**Comment by dcworldwide** [comment]
@stoffern Ok I've tested out your filters by porting your changes into my repo and using my test test harness. I've experiencing the same issues as I did previously.
- Sometimes the GL Surface is rendered as a black screen
- The filters, mostly don't work. Quite often you'll see filters with a pink hue. 

I can't help but feel there are some issues with React GL. 

Some examples, Amaro and Brannan

![screenshot_20160725-141008](https://cloud.githubusercontent.com/assets/4898751/17090348/314b850a-5273-11e6-90ba-7638a5b916f3.png)
![screenshot_20160725-141015](https://cloud.githubusercontent.com/assets/4898751/17090349/3154e8fc-5273-11e6-87e7-d5f7827e5508.png)
![Uploading Screenshot_20160725-142137.png…]()

**Comment by st0ffern** [comment]
I dont think this is a issue with big files as some filters only use small files
Could it be that we need to preload the images so they are ready for gl-react?

@gre any ideas what this can be?

**Comment by gre** [comment]
not sure about the "sometimes black" but it might be related to a buggy Android implementation of gl-react-native (see answered https://github.com/gre/gl-react-image/issues/3#issuecomment-235230040 )

**Comment by st0ffern** [comment]
@gre i dont think the render problems is related to the `gl-react-image`
i have used this https://github.com/stoffern/gl-react-instagramfilters in gl-react and the same problem appears here. i will try to investigate it futher..

**Comment by dcworldwide** [comment]
Ok, i've done some more testing. I've checked in my test harness changes. 
- Tested only android 
- I'm now loading all shader uniforms locally using require(). It appears more stable, however it still crashes intermittently. @gre how can I produce a stack trace for you?
- Upgraded to 2.28.3
- @stoffern I think now the issue is primarily with the fragment shader implementations. Some work (Hudson, Inkwell), the others that I've tested (Amaro, Brannan, S1977, Valencia) don't.  I don't understand glsl well at all so i'm stuck here. 

![screenshot_20160808-123951](https://cloud.githubusercontent.com/assets/4898751/17467783/3ec1dc0a-5d65-11e6-9504-ef6412fb98ed.png)

**Comment by st0ffern** [comment]
@dcworldwide i think we should move this issue over to `gl-react`
I have the same problem on web and on android. I think this is a issue with preloading images.
I will setup a test to preload images before launching gl-react-native to see if it works better then.
Can do this by the end of the week.

Please use web when you try to find errors with GLSL in gl-react, this is because android and possible ios just crashes when you get a error. on web you get a error message, and the error is handled correct :smile: 

cc @gre

**Comment by dcworldwide** [comment]
Hmm, is the following not the same as pre-loading textures?

```
const resolveAssetSource = require("react-native/Libraries/Image/resolveAssetSource");
const inputImageTexture2 = resolveAssetSource(require('../../assets/filters/brannanBlowout.png'))
```

I can cross-post the issue. I raised it here, because this project acts as a reproduction.

**Comment by st0ffern** [comment]
No, react-native uses Fresco lib to load images. that is a async call for opening stream.

It will:
1. load react-native
2. load fresco image
3. fresco opens stream in background and read image.
4. react-native continues to load
5. gl-react-native til load
6. fresco filestream read completes and sends image to react-native

small images may work as they are not to long to load. bigger images would possible fail to load before gl-react-native and gl-react

PS gl-react-image also uses fresco

**Comment by mersay** [comment]
android is crashing for me. can't even load anything

**Comment by dcworldwide** [comment]
@mersay  probably not related to this, but to a crashing shader. Makes sure you're uniforms are all in use and valid

**Comment by st0ffern** [comment]
@mersay not related to this. if it crashes please provide log details.
If ReactNative dont provide crash error, then react on web will.

@dcworldwide if the GLSurface renders black i guess you are using web og url images?
Make shure to preload them before rendering GLSurface..

**Comment by fakenickels** [comment]
I know this is a bit hacky but if you set the Surface wrapper an opacity of 0.1 the image shows correctly
